### DISCUSSION TITLE
Teilnahmebestätigung mit Gebühr und Übernachtungskosten

### DISCUSSION THREAD

**Discussion by tolot27** [openjverein/jverein#329] [Q&A]
Moin!

Wie kann ich denn eine Teilnahmebestätigung für eine Veranstaltung per E-Mail versenden, in der eine Teilnahmegebühr sowie Übernachtungskosten für ein Mitglied und einem oder mehreren Familienmitglieder aufgeführt werden? Basierend auf der Teilnahmebestätigung sollen die Mitglieder dann auch gleich überweisen können (unterstützt durch QR Code). Bei Lastschrift soll es gleich eine Pre-Notification enthalten und später soll die Lastschrift dann auch durchgeführt werden.

Ich dachte, ich könnte es über Zusatzbeiträge abbilden. Dann könnte ich einen Abrechnungslauf machen. Allerdings bekomme ich den Zusatzbeitrag nicht in eine Mailvorlage rein. Er wird in den Variablen der Mail-Empfänger nicht aufgeführt.
Über freie Formulare geht es auch nicht, da immer ein Formular angegeben werden muss, auch bei Ausgabe EMAIL. Ich will eigentlich keine Attachments erzeugen.

Bin für jeden Hinweis dankbar. :-)

**Comment by lenilsas** [comment]
Das kannst du über Lesefelder machen. Hier musst du dir ein entsprechendes Script schreiben mit dem du die Zusatzbeiträge bekommst. Dann kannst du dieses Lesefeld in der Mail verwenden.

**Reply by tolot27** [reply]
Okay, soweit verstanden und umgesetzt. Nur wie bekomme ich dann Zugriff auf das Mitglied-Objekt? Die Variablen der `MitgliedskontoMap` werden nämlich leider nicht mit bei den Mailvorlagen bereitgestellt. Da könnte ich nämlich ganz einfach auf `mitgliedskonto_summe_offen` zugreifen, wenn die Variable existieren würde.
Für die Zusatzbeträge gibt es auch keine Variablen.

Nebenbei würde ich auch gern eine Funktion schreiben, die die Mitgliedsnummern einer Familie auflistet. Dazu müsste ich jedoch auf den `FamilienbeitragNode` zugreifen können. Aber das ist off-topic für diese Fragestellung hier.

**Reply by lenilsas** [reply]
Auf die Zusatzbeträge kannst du so zugreifen:

> import de.jost_net.JVerein.rmi.Mitglied;
> import de.jost_net.JVerein.rmi.Zusatzbetrag;
> import de.willuhn.datasource.rmi.DBIterator;
> import de.jost_net.JVerein.Einstellungen;
> 
>     DBIterator it = Einstellungen.getDBService()
>         .createList(Zusatzbetrag.class);
>     it.addFilter("mitglied = " + mitglied_id);
>    String text = "";
>     while (it.hasNext())
>     {
>       Zusatzbetrag z = it.next();
>       text = text + "\n" + z.getBuchungstext();
>     }
> return text;
> 
Das als erster Ansatz, dann muss man noch prüfen welche Zusatzbeträge offen sind etc.
Dafür muss man sich mit dem Code von Jverein etwas auskennen....
Ich werde auch nochmal prüfen, ob man die Maps evtl. erweitern kann, so dass hier mehr Variablen verfügbar sind.

**Reply by tolot27** [reply]
D. h. ich kann auf die Objekte direkt zugreifen bzw. hole sie mir im gezeigten Beispiel direkt von der DB. Fantastisch!
Genial wäre es natürlich, wenn ein Mitgliedsobjekt auch in der Map als Referenz gespeichert wäre.

**Reply by tolot27** [reply]
Hm, eigentlich will ich ja nur die noch offenen Zusatzbeträge erfassen. Mit `isAktiv()` komme ich da nicht weiter, da ich bereits einen Abrechnungslauf gemacht habe und somit alles bereits als Sollbuchung (per Überweisung nur bei der ersten Zahlung) beim Mitgliedskonto eingetragen ist. Nur so kann ich es ja später beim automatischen Kontoabgleich auch automatisch zuordnen lassen.
Oder wie ist die Logik bei dir?

**Reply by tolot27** [reply]
> Oder wie ist die Logik bei dir?

Einfacher ist es, vor dem Abrechnungslauf die Mail zu versenden. 😄
